core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -2000,6 +2025,149 @@ private[spark] class DAGScheduler(
     }
   }
 
+  /**
+   * Schedules shuffle merge finalize.
+   */
+  private[scheduler] def scheduleShuffleMergeFinalize(stage: ShuffleMapStage): Unit = {
+    logInfo(("%s (%s) scheduled for finalizing" +
+      " shuffle merge in %s s").format(stage, stage.name, shuffleMergeFinalizeWaitSec))
+    shuffleMergeFinalizeScheduler.schedule(
+      new Runnable {
+        override def run(): Unit = finalizeShuffleMerge(stage)
+      },
+      shuffleMergeFinalizeWaitSec,
+      TimeUnit.SECONDS
+    )
+  }
+
+  /**
+   * DAGScheduler notifies all the remote shuffle services chosen to serve shuffle merge request for
+   * the given shuffle map stage to finalize the shuffle merge process for this shuffle. This is
+   * invoked in a separate thread to reduce the impact on the DAGScheduler main thread, as the
+   * scheduler might need to talk to 1000s of shuffle services to finalize shuffle merge.
+   */
+  private[scheduler] def finalizeShuffleMerge(stage: ShuffleMapStage): Unit = {
+    logInfo("%s (%s) finalizing the shuffle merge".format(stage, stage.name))
+    externalShuffleClient.foreach { shuffleClient =>
+      val shuffleId = stage.shuffleDep.shuffleId
+      val numMergers = stage.shuffleDep.getMergerLocs.length
+      val numResponses = new AtomicInteger()
+      val results = (0 until numMergers).map(_ => SettableFuture.create[Boolean]())
+      val timedOut = new AtomicBoolean()
```

Comment:
`timedOut` is insufficient to prevent a race - we can have:
a) `ShuffleMergeFinalized` being posted in IO thread when all partitions complete.
b) Futures timeout before `increaseAndCheckResponseCount` completes.

This will result in two `ShuffleMergeFinalized` being posted - and the second one will cause invalidation of merge result.

To fix this, I propose the following:

a) Rename `timedOut` to `shouldPostFinalizedEvent` - with default value of `true`.
b) post the event only when `shouldPostFinalizedEvent.getAndSet(false)` returns true (in `increaseAndCheckResponseCount` and in `TimeoutException` exception catch block).

This will ensure that there is only one `ShuffleMergeFinalized` posted - either due to timeout or due to required mergers completing.

Thoughts ?

Reply:
After having offline discussion, we simplified it further just by posting `ShuffleMergeFinalized` only in the finally block that way it guarantees the event itself would get posted only once.